test/e2e/template/helmfile/tmpl_test.go

```diff
@@ -0,0 +1,254 @@
+		wantErr:    true,
+	},
+}
+
+var toYamlTestCases = []tmplTestCase{
+	{
+		data: map[string]string{
+			"test": "test",
+		},
+		name:       "toYaml",
+		tmplString: `{{ . | toYaml }}`,
+		output:     "test: test\n",
+	},
+}
+
+var fromYamlTestCases = []tmplTestCase{
+	{
+		name: "fromYaml",
+		tmplString: `{{ $value :=  "test: test" | fromYaml }}
+		             {{- $value.test }}`,
+		output: "test",
+	},
+}
+
+var setValueAtPathTestCases = []tmplTestCase{
+	{
+		data: map[string]interface{}{
+			"root": map[string]interface{}{
+				"testKey": map[string]interface{}{
+					"testKey2": "test",
+				},
+			},
+		},
+		name: "setValueAtPath",
+		tmplString: `{{- $newValues := . | setValueAtPath "root.testKey.testKey2" "testNewValue" }}
+		             {{- $newValues.root.testKey.testKey2 }}`,
+		output: "testNewValue",
+	},
+	{
+		data: map[string]interface{}{
+			"root": "test",
+		},
+		wantErr:    true,
+		name:       "setValueAtPathWithInvalidPath",
+		tmplString: `{{ . | setValueAtPath "root.nokey" "testNewValue" }}`,
+	},
+}
+
+var getTestCases = []tmplTestCase{
+	{
+		data: map[string]interface{}{
+			"root": map[string]interface{}{
+				"testGetKey": map[string]interface{}{
+					"testGetKey2": "test",
+				},
+			},
+		},
+		name:       "get",
+		tmplString: `{{- . | get "root.testGetKey.testGetKey2" "notfound" }}`,
+		output:     "test",
+	},
+	{
+		data: map[string]interface{}{
+			"root": map[string]interface{}{
+				"testGetKey": map[string]interface{}{
+					"testGetKey2": "test",
+				},
+			},
+		},
+		name:       "getNotExistWithDefault",
+		tmplString: `{{- . | get "root.testGetKey.testGetKey3" "notfound" }}`,
+		output:     "notfound",
+	},
+}
+
+var tplTestCases = []tmplTestCase{
+	{
+		data: map[string]interface{}{
+			"root": "tplvalue",
+		},
+		name:       "tpl",
+		tmplString: `{{ . | tpl "{{ .root }}" }}`,
+		output:     "tplvalue",
+	},
+	{
+		data:       map[string]interface{}{},
+		name:       "tplInvalidTemplate",
+		wantErr:    true,
+		tmplString: `{{ . | tpl "{{ .root }}" }}`,
+	},
+}
+
+// tmplTestCases are the test cases for the template tests
+type tmplE2e struct {
+	tcs []tmplTestCase
+}
+
+// append for append testcase into tmplTestCases
+func (t *tmplE2e) append(ts ...tmplTestCase) {
+	t.tcs = append(t.tcs, ts...)
+}
+
+// load for  load testcase into tmplTestCases
+func (t *tmplE2e) load() {
+	t.append(requireEnvTestCases...)
+	t.append(requiredTestCases...)
+	t.append(envExecTestCases...)
+	t.append(execTestCases...)
+	t.append(readFileTestCases...)
+	t.append(toYamlTestCases...)
+	t.append(fromYamlTestCases...)
+	t.append(setValueAtPathTestCases...)
+	t.append(getTestCases...)
+	t.append(tplTestCases...)
+}
+
+var tmplE2eTest = tmplE2e{}
+
+// TestTmplStrings tests the template string
+func TestTmplStrings(t *testing.T) {
+	c := &tmpl.Context{}
+	c.SetBasePath(".")
+	c.SetReadFile(os.ReadFile)
+	tmpl := template.New("stringTemplateTest").Funcs(c.CreateFuncMap())
+
+	tmplE2eTest.load()
```

Comment:
Sorry for very nit-picky again but I personally prefer it being an inlined sequence of `tcs = append(tcs, requireEnvTestCases...)` without the indirection(extra struct and its instance plus another two functions to just compose an array of test cases).

Reply:
I will continuous optimization. Thanks very much. @mumoshu